mypy/checkexpr.py

```diff
@@ -1238,7 +1238,9 @@ def infer_overload_return_type(self,
         for typ in plausible_targets:
             overload_messages = self.msg.clean_copy()
             prev_messages = self.msg
+            assert self.msg is self.chk.msg
```

Comment:
I think there's a slightly different pattern to ignore messages during a certain block of code, see e.g. checkmember.py L183. I'll test your code next.